src/elements/reloc_section.rs

```diff
@@ -0,0 +1,244 @@
+use std::io::{Read, Write};
+
+use super::{CountedList, CountedListWriter, CountedWriter, Deserialize, Error, Serialize, VarInt32, VarUint32, VarUint7};
+
+const R_WEBASSEMBLY_FUNCTION_INDEX_LEB: u8 = 0;
+const R_WEBASSEMBLY_TABLE_INDEX_SLEB: u8 = 1;
+const R_WEBASSEMBLY_TABLE_INDEX_I32: u8 = 2;
+const R_WEBASSEMBLY_MEMORY_ADDR_LEB: u8 = 3;
+const R_WEBASSEMBLY_MEMORY_ADDR_SLEB: u8 = 4;
+const R_WEBASSEMBLY_MEMORY_ADDR_I32: u8 = 5;
+const R_WEBASSEMBLY_TYPE_INDEX_LEB: u8 = 6;
+const R_WEBASSEMBLY_GLOBAL_INDEX_LEB: u8 = 7;
+
+/// Relocation information.
+#[derive(Clone, Debug)]
+pub struct RelocSection {
+	// Name of this section.
+	name: String,
+
+	// ID of the section containing the relocations described in this section.
+	section_id: u32,
+
+	// Name of the section containing the relocations described in this section. Only set if section_id is 0.
+	relocation_section_name: Option<String>,
+
+	// Relocation entries.
+	entries: Vec<RelocationEntry>,
+}
+
+impl RelocSection {
+	pub fn deserialize<R: Read>(
+		name: String,
+		rdr: &mut R,
+	) -> Result<Self, Error> {
+		let section_id = VarUint32::deserialize(rdr)?.into();
+		let relocation_section_name =
+			if section_id == 0 {
+				Some(String::deserialize(rdr)?)
+			}
+			else {
+				None
+			};
+
+		let entries = CountedList::deserialize(rdr)?.into_inner();
+
+		Ok(RelocSection {
+			name,
+			section_id,
+			relocation_section_name,
+			entries,
+		})
+	}
+}
+
+impl Serialize for RelocSection {
+	type Error = Error;
+
+	fn serialize<W: Write>(self, wtr: &mut W) -> Result<(), Error> {
+		let mut counted_writer = CountedWriter::new(wtr);
+		self.name.serialize(&mut counted_writer)?;
+
+		VarUint32::from(self.section_id).serialize(&mut counted_writer)?;
+
+		if let Some(relocation_section_name) = self.relocation_section_name {
+			relocation_section_name.serialize(&mut counted_writer)?;
+		}
+
+		let counted_list = CountedListWriter(self.entries.len(), self.entries.into_iter());
+		counted_list.serialize(&mut counted_writer)?;
+
+		counted_writer.done()?;
+
+		Ok(())
+	}
+}
+
+/// Relocation entry.
+#[derive(Clone, Debug)]
+pub enum RelocationEntry {
+	// Function index.
+	FunctionIndexLeb {
+		offset: u32,
```

Comment:
Also there should be doc comment about each field in every variant entry

Reply:
Done.